ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/storage/StorageProvider.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.storage;
+
+import tech.pegasys.pantheon.ethereum.db.BlockchainStorage;
+import tech.pegasys.pantheon.ethereum.mainnet.ProtocolSchedule;
+import tech.pegasys.pantheon.ethereum.worldstate.WorldStateStorage;
+
+import java.io.Closeable;
+
+public interface StorageProvider extends Closeable {
+
+  BlockchainStorage createBlockchainStorage(ProtocolSchedule<?> protocolSchedule);
```

Comment:
Can we do this without the ProtocolSchedule argument?  Is there a case where we don't know the protocol schedule when we are making the decisions about what kind of blockchain storage we provide?  We could just curry a function taking that into the constructor and make it a provider.  It's a bit of work on our end but when we go to DI the framework will wire it together.

Reply:
Unfortunately we don't have the `ProtocolSchedule` available at the time we want to create the `StorageProvider`.  The issue is that you need a `BlockHashFunction` to load any blocks from storage and that depends on the `ProtocolSchedule`.  We very likely want to remove that dependency which would mean removing the `hash` function from `Block` and `BlockHeader` - anyone wanting the hash would have to get the `BlockHashFunction` from the `ProtocolSchedule` and pass the `BlockHeader` to it.  That's not a simple change to make and we would need to preserve the hash caching that `BlockHeader` currently provides internally as it's a fairly expensive operation and is performed quite a bit.

Reply:
Yes, it's hard to go halfway with a DI style without a framework.  A lot of these issues would melt into the module (one for each consensus protocol) but without that framework we can't go halfway.  So since we don't have a forma DI framework handling the injections for us this is the engineering compromise we will need to do.